plugins/Compressor/CompressorControlDialog.cpp

```diff
@@ -0,0 +1,753 @@
+		p.fillRect(m_kneeWindowSizeX + 1, 0, m_windowSizeX, m_kneeWindowSizeY, QColor("transparent"));
+		p.setCompositionMode(QPainter::CompositionMode_SourceOver);
+
+		p.end();
+
+		p.begin(&m_kneePixmap2);
+		
+		p.setCompositionMode(QPainter::CompositionMode_Source);
+		p.fillRect(0, 0, m_windowSizeX, m_kneeWindowSizeY, QColor("transparent"));
+		p.setCompositionMode(QPainter::CompositionMode_SourceOver);
+
+		p.end();
+
+		m_lastKneePoint = 0;
+	}
+
+	// Start drawing second knee layer
+	p.begin(&m_kneePixmap2);
+
+	p.setRenderHint(QPainter::Antialiasing, false);
+
+	int kneePoint = dbfsToXPoint(ampToDbfs(peakAvg));
+	if (kneePoint > m_lastKneePoint)
+	{
+		QRectF knee2Rect = QRect(m_lastKneePoint, 0, kneePoint - m_lastKneePoint, m_kneeWindowSizeY);
+		p.drawPixmap(knee2Rect, m_kneePixmap, knee2Rect);
+	}
+	else
+	{
+		p.setCompositionMode(QPainter::CompositionMode_Source);
+		p.fillRect(kneePoint, 0, m_lastKneePoint, m_kneeWindowSizeY, QColor("transparent"));
+		p.setCompositionMode(QPainter::CompositionMode_SourceOver);
+	}
+	m_lastKneePoint = kneePoint;
+
+	p.end();
+
+	if (m_controls->m_effect->m_redrawThreshold)
+	{
+		p.begin(&m_miscPixmap);
+
+		p.setCompositionMode(QPainter::CompositionMode_Source);
+		p.fillRect(0, 0, m_windowSizeX, m_windowSizeY, QColor("transparent"));
+		p.setCompositionMode(QPainter::CompositionMode_SourceOver);
+
+		p.setRenderHint(QPainter::Antialiasing, true);
+
+		// Draw threshold lines
+		p.setPen(QPen(m_threshColor, 2, Qt::DotLine));
+		p.drawLine(0, threshYPoint, m_windowSizeX, threshYPoint);
+		p.drawLine(threshXPoint, 0, threshXPoint, m_kneeWindowSizeY);
+
+		p.end();
+
+		m_controls->m_effect->m_redrawThreshold = false;
+	}
+
+	m_lastPoint = yPoint;
+	m_lastGainPoint = yGainPoint;
+
+	update();
+}
+
+
+
+void CompressorControlDialog::paintEvent(QPaintEvent *event)
+{
+	if (isVisible())
+	{
+		QPainter p(this);
+
+		p.setCompositionMode(QPainter::CompositionMode_Source);
+		p.fillRect(0, 0, m_windowSizeX, m_windowSizeY, QColor("transparent"));
+		p.setCompositionMode(QPainter::CompositionMode_SourceOver);
+
+		p.drawPixmap(0, 0, m_graphPixmap);
+		p.drawPixmap(0, 0, m_visPixmap);
+		p.setOpacity(0.25);
+		p.drawPixmap(0, 0, m_kneePixmap);
+		p.setOpacity(1);
+		if (m_controls->m_effect->isEnabled() && m_controls->m_effect->isRunning())
+		{
+			p.drawPixmap(0, 0, m_kneePixmap2);
+		}
+		p.drawPixmap(0, 0, m_miscPixmap);
+
+		p.end();
+	}
+}
+
+
+inline int CompressorControlDialog::dbfsToYPoint(float inDbfs)
+{
+	return (-((inDbfs + m_dbRange) / m_dbRange) + 1) * m_windowSizeY;
+}
+
+inline int CompressorControlDialog::dbfsToXPoint(float inDbfs)
+{
+	return m_kneeWindowSizeY - dbfsToYPoint(inDbfs);
+}
+
+
+void CompressorControlDialog::resizeEvent(QResizeEvent *event)
+{
+	resetCompressorView();
+}
+
+
+void CompressorControlDialog::wheelEvent(QWheelEvent * event)
+{
+	const float temp = m_dbRange;
+	m_dbRange = round(qBound(3.f, m_dbRange - event->delta() / 20.f, 96.f) / 3.f) * 3.f;
```

Comment:
A comment about where the constants come from can help in case wheel event deltas change or cause issues on other systems. What defines the bounds? Why 3? Are any of these a target for `SOME_CONSTANT_VARIABLE_NAME`? etc.

Reply:
It's 3 db per grid line.  I'll add a const variable for that.

Reply:
Oh, thanks for catching that issue.  The formula was a temporary lazy formula that just worked, not something I intended to keep there.  Something as simple as somebody using a different mouse or something could potentially mess with this formula in undesirable ways I think.  I'll fix that.